plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/embeddings/NameEmbeddings.kt

```diff
@@ -9,151 +9,51 @@ import org.jetbrains.kotlin.fir.symbols.impl.FirConstructorSymbol
-}
-
-data class GlobalFunctionName(override val packageName: FqName, val name: Name, val typeSignature: String) : FqMangledName {
-    override val postfix: String = "global$${name.asStringStripSpecialMarkers()}$${typeSignature}"
-}
-
-/**
- * Mangle a type's name into an identifier supported by Viper.
- * When the type is Nullable, its name will be prefixed by `NT` (Nullable Type).
- * Otherwise, the prefix will be `T` (Type).
- */
-val ConeKotlinType.mangledName: String
-    get() {
-        val prefix = when (this.isNullable) {
-            true -> "NT" // Nullable Type
-            false -> "T" // Type
-        }
-        // Check if there are any type argument in type's name
-        val mangled = toString()
-            .replace("/", "_")
-            .replace(Regex("[? ]"), "")
-            .replace(Regex("[<>]"), "__")
-            .replace(",", "$")
-        return prefix + mangled
+fun CallableId.embedScopeName(): NameScope =
+    when (val id = this.classId) {
+        null -> GlobalScope(packageName)
+        else -> ClassScope(packageName, id.embedName())
     }
 
-/**
- * Create a `String` embedding the type names, joined together by a separator character (`$`).
- * This function is useful to handle cases with function overloading.
- */
-fun FirFunctionSymbol<*>.embedTypeSignature(): String = valueParameterSymbols.joinToString("$") { param ->
-    param.resolvedReturnType.mangledName
-}
+fun CallableId.embedScoped(f: CallableId.() -> KotlinName) = ScopedKotlinName(embedScopeName(), f())
+fun CallableId.embedScopedWithType(type: TypeEmbedding, f: CallableId.() -> KotlinName) = ScopedKotlinName(embedScopeName(), f(), type)
 
-// Parameters always have scope depth equal to zero
-fun FirValueParameterSymbol.embedName(): LocalName = LocalName(name, 0)
+fun ClassId.embedName(): ScopedKotlinName = ScopedKotlinName(GlobalScope(packageFqName), ClassKotlinName(shortClassName))
+fun CallableId.embedGetterName(): ScopedKotlinName = embedScoped { GetterKotlinName(callableName) }
+fun CallableId.embedSetterName(): ScopedKotlinName = embedScoped { SetterKotlinName(callableName) }
+fun CallableId.embedMemberPropertyName(): MangledName = embedScoped { MemberKotlinName(callableName) }
 
-fun FirPropertyAccessorSymbol.embedName(): MangledName {
-    val className = propertySymbol.callableId.classId!!.embedName()
-    val propertyName = propertySymbol.callableId.callableName
-    return when {
-        isGetter -> ClassMemberGetter(className, propertyName)
-        isSetter -> ClassMemberSetter(className, propertyName)
-        else -> throw IllegalStateException("A property accessor must be a setter or a getter!")
-    }
+fun CallableId.embedPropertyName(scope: Int): MangledName = when {
+    isLocal -> ScopedKotlinName(LocalScope(scope), SimpleKotlinName(callableName))
+    className != null -> embedMemberPropertyName()
+    else -> throw Exception("Name is neither local nor bound to a class; we do not know how to handle this.")
```

Comment:
I am a bit against in throwing the most generic exception in Kotlin. Maybe an `UnsupportedOperationException` would be better, check [here](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin/-unsupported-operation-exception/).

Reply:
Thanks; `IllegalStateException` seems to make sense given it's what we use for the getter/setter not being identified.